core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -278,9 +278,24 @@ private[deploy] class Master(
       execOption match {
         case Some(exec) => {
           val appInfo = idToApp(appId)
+          val oldState = exec.state
           exec.state = state
-          if (state == ExecutorState.RUNNING) { appInfo.resetRetryCount() }
+
+          if (state == ExecutorState.RUNNING) {
+            assert(oldState == ExecutorState.LOADING,
+              s"executor $execId state transfer from $oldState to RUNNING is illegal")
+            appInfo.resetRetryCount()
+          }
+
           exec.application.driver.send(ExecutorUpdated(execId, state, message, exitStatus))
+
+          if (state == ExecutorState.LOADING) {
+            assert(oldState == ExecutorState.LAUNCHING,
+              s"executor $execId state transfer from $oldState to LOADING is illegal")
+            exec.application.driver.send(
```

Comment:
Yes,  driver only can be told when master get executor state from worker, that is a correct sequence for state transition.